base/iterators.jl

```diff
@@ -1602,4 +1602,83 @@ end
 # be the same as the keys, so this is a valid optimization (see #51631)
 pairs(s::AbstractString) = IterableStatePairs(s)
 
+"""
+    nth(itr, n::Integer)
+
+Get the `n`th element of an iterable collection. Throw a `BoundsError`[@ref] if not existing.
+Will advance any `Stateful`[@ref] iterator.
+
+See also: [`first`](@ref), [`last`](@ref), [`nth`](@ref)
+
+# Examples
+```jldoctest
+julia> Iterators.nth(2:2:10, 4)
+8
+
+julia> Iterators.nth(reshape(1:30, (5,6)), 6)
+6
+
+julia> stateful = Iterators.Stateful(1:10); nth(stateful, 7)
+7
+
+julia> first(stateful)
+8
+```
+"""
+nth(itr, n::Integer) = _nth(itr, n)
+
+# Count
+nth(itr::Count, n::Integer) = n > 0 ? itr.start + itr.step * (n - 1) : throw(ArgumentError("n must be positive."))
+# Repeated
+nth(itr::Repeated, ::Integer) = itr.x
+# Take(Repeated)
+nth(itr::Take{<:Repeated}, n::Integer) =
+    n > itr.n ? throw(BoundsError(itr, n)) : nth(itr.xs, n)
+
+# infinite cycle
+function nth(itr::Cycle{I}, n::Integer) where {I}
+    if IteratorSize(I) isa Union{HasShape, HasLength}
+        _nth(itr.xs, mod1(n, length(itr.xs)))
+    else
+        _nth(itr, n)
+    end
+end
+
+# finite cycle: in reality a Flatten{Take{Repeated{O}}} iterator
+function nth(itr::Flatten{Take{Repeated{O}}}, n::Integer) where {O}
+    if IteratorSize(O) isa Union{HasShape, HasLength}
+        cycles = itr.it.n
+        repeated = itr.it.xs.x
+        k = length(repeated)
+        n > k*cycles ? throw(BoundsError(itr, n)) : _nth(repeated, mod1(n, k))
+    else
+        _nth(itr, n)
+    end
+end
+
+@inline _nth(itr, n) = first(drop(itr, n-1))
+@inline _nth(itr::AbstractArray, n) = itr[begin + n-1]
+"""
+    nth(n::Integer)
+
+Return a function that gets the `n`-th element from any iterator passed to it.
+Throw a `BoundsError`[@ref] if not existing.
```

Comment:
I would delete this `nth(n)` function.   We don't have that for `getindex` or any similar function, so I don't see why we should have it hear.  People can always just write `itr -> nth(itr, n)` if they need this.

Reply:
I'm following the spec that came out of triage, which includes the single argument version.
At least for my understanding of Jeff and Lilith comments.

Reply:
One vote that `nth(n)` might be handy. 

But I'd shorten the docstring not to include all the same disclaimers.. it needs to say this is like `xs -> nth(xs, n)` and not much more. Refer to main docstring for Stateful and BoundsError etc. 

`map(Iterators.nth(3), my_vec)` also won't work in a `jldoctest`. I suggest:
```
julia> map(fifth_element, ("Willis", "Jovovich", "Oldman"))
('i', 'v', 'a')
```